### PR TITLE
Seats paging and status filter

### PR DESCRIPTION
**Pull request by dagbay-rh**
# Summary

## What
* Add a new `status` param on `Get /seats` endpoint
* Pass along the status param to the ams query to reduce the result set from ams (potentially saves bandwidth)
* Deprecate the `excludeStatus` param on the same endpoint
* EDIT: renamed `ams/test_client` to `ams/mock_client` since I added a `ams/client_test` file, to avoid confusion. I also changed the struct from `TestClient` to `Mock`
* EDIT: renamed `controllers/ams` and `controllers/seat_test` to `seats` and `seats_test` respectively. Trying to standardize some of the naming we used here

## Why
Right now paging does not work properly if you use `excludeStatus` on a `Get /seats` call and the result set is larger that a page size. This is because we include paging size and offset in the call to ams, but then filter out certain seats on the result set without updating the paging links. This results in scenarios where the result set from ams can be greater than the actual payload data returned by the endpoint, which breaks the paging.

## Example
Given: you have 13 total seats for a user, with 10 in `Deprovisioned` status and 3 in `Active` status

When: you hit `GET /api/entitlements/v1/seats?excludeStatus=Deprovisioned&page=10&offset=0` (to see active seats)

Then:
* entitlements queries ams for all seats for the user, with page size set to 10 and offset to 0
* ams returns the first 10/13 seats, in order of when they were created
* the first 10 seats are now all in `Deprovisioned` status
* entitlements removes the 10 seats with `Deprovisioned` status from the result payload data, without updating page links
* The user sees a response like this:
```json
{
    "allowed": 50,
    "consumed": 3,
    "data": [],
    "links": {
        "first": "/api/entitlements/v1/seats/?limit=10&offset=0",
        "next": "/api/entitlements/v1/seats/?limit=10&offset=10",
        "previous": "/api/entitlements/v1/seats/?limit=10&offset=0"
    },
    "meta": {
        "count": 0
    }
}
```
* this means the first page is "empty", and the user has to know to query for the next page to actually see their results

This is bad ux, since they could potentially have to parse thru multiple pages before actually getting the data they requested, with no way of knowing which page will not be empty. There is also no guarantee all the empty pages will be listed first, meaning they could potentially have to parse thru multiple pages looking for seats, and having to count up to their requested page size to actually display what they asked for.

This is also a potential performance problem, since we could eventually have 100s of seats deprovisioned, meaning our calls to ams could blow up quite fast because we would be calling them multiple times even just to get the data the user wants.

So, overall I think the `status` param is easier to understand, and allows to to pass the status in our query to ams, potentially reducing the amount of data returned over the wire as well as number of calls to ams

## Tickets
* https://issues.redhat.com/browse/RHCLOUD-27842

# Platform Security

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist 

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

**Review by Ellen-Yi-Dong — APPROVED** [review]
LGTM!

File: ams/client.go [hunk 1]
```diff
@@ -149,18 +150,26 @@ func (c *Client) GetSubscription(subscriptionId string) (*v1.Subscription, error
...
 		return nil, err
 	}
-	q := "plan.id LIKE 'AnsibleWisdom'"
-	q += " AND "
-	q += fmt.Sprintf("organization_id = '%s'", amsOrgId)
+
+	queryBuilder := NewQueryBuilder().
+		Like("plan.id", "AnsibleWisdom").
+		And().
+		Equals("organization_id", amsOrgId)
```
Comment (jhjaggars) [inline ams/client.go:162]: Since this is being interpolated into a string, ensure that the value is safe

Comment (jhjaggars) [reply ams/client.go:162]: Also, I appreciate that it's been unvalidated in the past, sorry for that :)

File: ams/client.go [hunk 2]
```diff
@@ -149,18 +150,26 @@ func (c *Client) GetSubscription(subscriptionId string) (*v1.Subscription, error
...
+
+	if statuses != nil && len(statuses) > 0 {
+		queryBuilder = queryBuilder.And().In("status", statuses)
```
[the remark this answers is not in the document]
Comment (dagbay-rh) [reply ams/client.go:165]: heck yeah, I'll validate them against the valid values from our spec

File: ams/client.go (hunk 1 above)
Comment (jhjaggars) [reply ams/client.go:165]: Maybe allowlist something like this `/[a-zA-Z0-9]+/`.

Comment (jhjaggars) [reply ams/client.go:165]: Though, now that you remind me maybe it's not really helpful.

Comment (dagbay-rh) [reply ams/client.go:166]: sounds good :rocket: updated

Comment (dagbay-rh) [reply ams/client.go:166]: I also validated the user org id from the user request, since we also query ams with that

**Review by jhjaggars — APPROVED** [review]
looks awesome!

**Pull request merged**